crates/dbsp/src/operator/dynamic/input_upsert.rs

```diff
@@ -556,11 +567,79 @@ where
 
         self.input_batch_stats.add_batch(n_updates);
 
+        // # Overview
+        //
+        // We divide processing upserts for a given key in our input stream into
+        // two parts:
+        //
+        // 1. Inserting the key's new value.
+        //
+        // 2. Deleting the key's old value (a retraction).
+        //
+        // A truly new key has only #1, a key that existed and is being deleted
+        // has only #2, a key that existed and is being inserted has both #1 and
+        // #2.  A key that existed and is being inserted with the same value is
+        // a no-op and we have to make it cancel out one way or another.
```

Comment:
what does "one way or another" mean?
You can either produce an empty Z-set, or a reduce this case to the second case - two values that happen to cancel each other.